### ISSUE TITLE
LEDE Issue tracking

### ISSUE THREAD

**Issue by riptidewave93** [riptidewave93/LEDE-APU2#2]
## This is an issue created to track what has been tested, and confirmed working on this new fork for LEDE.

**Confirmed Working:**
- ath10k & ath9k WiFi
- CPU Scaling (default via upstream)
- SD Card & SSD Booting 
- LEDs/Board Detection/Reset Button
- NCT5104D GPIO Driver
- flashrom (BIOS upgrade support)
- SP5100 Watchdog
- PC Buzzer
- AES-NI acceleration (added in kernel)
- lm-sensors

**Not Working Yet:**
- Nothing - You tell me?

**Patches to Submit**
- Device Profile - https://github.com/lede-project/source/pull/883

**Patches Merged**
- sp5100_tco Kernel Patches - https://github.com/lede-project/source/commit/4fab08183b973728a5db2fc691f4845e6a48d0b4
- sp5100_tco Kernel Module - https://github.com/lede-project/source/commit/0d9f922230aa02821c538a87c1d10584cff6c61c
- Beep package - https://github.com/openwrt/packages/commit/99605fb63069befb79db8b0cc62f9520734b8263
- GPIO Kernel Support - https://github.com/lede-project/source/commit/e3271d545d8f29f42a79d7cd5595858e67eb95e7
- GPIO Driver (as package) - https://github.com/lede-project/source/commit/a5eee54fd3691732885f0c80ecbf4530ff78d03f
- LED Driver (as package) - https://github.com/lede-project/source/commit/fe12d51cd49ddad71b16839eef10f9d307241e7a
- Userspace Support - https://github.com/lede-project/source/commit/9a6c9325564439e0c9bb3209931787a492fa8f9b

**Comment by CRCinAU** [comment]
So I'm going to be a pain..... Why don't you want to upstream stuff? :)

**Comment by riptidewave93** [comment]
@CRCinAU Main reason is the kernel patches that are added. Manually updating the revision in the build script ensures that things stay up to date. :)

EDIT: Derp, thought you were asking why I set the revision in build. (Sorry, I have not had coffee yet).

To PROPERLY answer your question, the big reason is the userspace & x64 kernel config changes. There are some things enabled (ex, LED driver, AES-NI) that will not work on other x64 platforms and previous discussions with the OpenWRT devs (before LEDE) noted that they didn't want a million different target devices for x86. While some of this could get merged up, things like system integration (reset button, LEDs, yada yada) would not be available.

**Comment by riptidewave93** [comment]
Just an update per @CRCinAU's comments. IRC conversations were had, and a plan was made to get this board upstream under the x86_64 subtarget as a device profile.

**Comment by riptidewave93** [comment]
@CRCinAU and @Strykar - http://lists.infradead.org/pipermail/lede-dev/2016-October/003326.html Initial RFC up for review.

**Comment by Strykar** [comment]
I'm not on the LEDE mailing list yet, joining now.

I should probably open a new issue for this but I'll wait till we move to an APU profile under LEDE/x64.
lm-sensors does not report individual core temps. The chip definitely supports it as FreeBSD/PFsense reports all 4 core temps correctly.

```
root@lede:/# sensors
k10temp-pci-00c3
Adapter: PCI adapter
temp1:        +62.0°C  (high = +70.0°C)
                       (crit = +105.0°C, hyst = +104.0°C)

root@lede:/# sensors -u
k10temp-pci-00c3
Adapter: PCI adapter
temp1:
  temp1_input: 62.375
  temp1_max: 70.000
  temp1_crit: 105.000
  temp1_crit_hyst: 104.000
```

**Comment by Gr4ffy** [comment]
It looks like there is a BusyBox hwclock in default LEDE build.

```
root@apu2:~# whereis hwclock
hwclock: /usr/sbin/hwclock /sbin/hwclock
root@apu2:~# /usr/sbin/hwclock -V
hwclock from util-linux 2.28
root@apu2:~# /sbin/hwclock -V
/sbin/hwclock: unrecognized option: V
BusyBox v1.25.0 () multi-call binary.

Usage: hwclock [-r] [-s] [-w] [-t] [-l] [-u] [-f FILE]

Query and set hardware clock (RTC)

    -r  Show hardware clock time
    -s  Set system time from hardware clock
    -w  Set hardware clock from system time
    -t  Set in-kernel timezone, correct system time
        if hardware clock is in local time
    -u  Assume hardware clock is kept in UTC
    -l  Assume hardware clock is kept in local time
    -f FILE Use specified device (e.g. /dev/rtc2)
```

**Comment by riptidewave93** [comment]
@Gr4ffy thanks for the report, the extra package was removed and busybox will be relied on. If you want to review the current changes, branch "lede-upstream-port" is currently synced with my working tree for RFC v2 (which will be posted after 7~ days from the post date of the first RFC)

**Comment by Gr4ffy** [comment]
Something is wrong with leds/gpio.

```
root@lede:~# logread | grep gpio-
Wed Oct 19 15:31:10 2016 kern.info kernel: [    2.480173] gpio-apu2: load APU2/LED GPIO driver module
Wed Oct 19 15:31:10 2016 kern.err kernel: [    2.485501] gpio-apu2: APU2 board not detected! Found: (null) (null)
Wed Oct 19 15:31:10 2016 kern.warn kernel: [    2.491882] gpio-apu2: probe of gpio-apu2 failed with error -13
Wed Oct 19 15:31:10 2016 kern.info kernel: [    2.497839] gpio-apu2: APU2 GPIO/LED driver module loaded
Wed Oct 19 15:31:10 2016 kern.err kernel: [    2.673112] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    2.722856] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    2.862426] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    2.870319] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    6.891286] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    6.959179] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    7.028997] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    7.076697] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.err kernel: [    8.663310] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
Wed Oct 19 15:31:10 2016 kern.info kernel: [    8.665122] gpio-nct5104d: Found nct5104d at 0x2e chip id 0xc452
Wed Oct 19 15:31:10 2016 kern.info kernel: [    8.665215] gpio-nct5104d: platform_driver_register
Wed Oct 19 15:31:10 2016 kern.info kernel: [    8.665398] gpio-nct5104d: Device added
Wed Oct 19 15:31:10 2016 kern.err kernel: [    8.686601] gpio-keys-polled gpio-keys-polled: unable to claim gpio 508, err=-517
```

**Comment by riptidewave93** [comment]
@Gr4ffy What BIOS version are you on? Also, can you share the outputs of the following?

```
cat /sys/devices/virtual/dmi/id/sys_vendor
cat /sys/devices/virtual/dmi/id/board_vendor
cat /sys/devices/virtual/dmi/id/product_name
cat /sys/devices/virtual/dmi/id/board_name
```

**Comment by Gr4ffy** [comment]
Reboot (HEAD, r1928) / lede-upstream-port

```
BIOS Information
        Vendor: coreboot
        Version: 88a4f96
        Release Date: 03/07/2016
root@lede:~# cat /sys/devices/virtual/dmi/id/sys_vendor
PC Engines
root@lede:~# cat /sys/devices/virtual/dmi/id/board_vendor
PC Engines
root@lede:~# cat /sys/devices/virtual/dmi/id/product_name
apu2
root@lede:~# cat /sys/devices/virtual/dmi/id/board_name
apu2
root@lede:~#
```

**Comment by riptidewave93** [comment]
@Gr4ffy Thanks, I was able to track down the issue. TL;DR the current APU2 BIOS reports as the "apu2" in SMBIOS, while the latest git head for coreboot reports as "APU2", which is what my board uses. Will get this fixed up and please keep the bugs coming!

EDIT: Fix merged into tracking branch/PR

**Comment by Gr4ffy** [comment]
Now it's ok.

**Comment by Strykar** [comment]
@riptidewave93 When do you suppose the first version with the APU2 as a profile as discussed might be ready to test?

**Comment by riptidewave93** [comment]
@Strykar you can build off of my "lede-upstream-port" branch if you want to test out the current progress of the port to upstream.

**Comment by Gr4ffy** [comment]
Today's lede-upstream-port build  "Reboot (HEAD, r2166)"
LED's driver is missing.
#10  CONFIG_LEDS_GPIO=y ?

Fixed

```
root@lede:~# dmesg | grep leds
[    2.495416] leds-apu2: load APU2/LED GPIO driver module
[    2.500810] leds-apu2: PCI Revision ID: 0x42
[    2.505191] leds-apu2: APU2 GPIO/LED driver module loaded
```

**Comment by Strykar** [comment]
Yes! LEDs working now (tested with `defaulton`), looking good so far on a few different builds.
I read your comment on the patches required and hence the hard git reset, any chance the kernel could be bumped to `4.4.29`?

**Comment by riptidewave93** [comment]
@Strykar If you are building from the "lede-upstream-port" branch, it should currently be on kernel 4.4.32.

**Comment by Strykar** [comment]
My mistake. Can the mini-PCIe 1 and mSATA LEDs on the board be controlled? I don't see them under `LED configuration`

**Comment by KyleSanderson** [comment]
Any update on upstreaming this?

**Comment by riptidewave93** [comment]
@Strykar Those are hardware LEDs that are not on GPIOs so they can not be controlled.

@KyleSanderson updates are in the PR at https://github.com/riptidewave93/LEDE-APU2/pull/5

**Comment by rotanid** [comment]
i'd like to add (as discussed with Chris via email) that although the APU2 is able to boot from SDCard, it isn't possible to reboot with most of the cards i tested - the system doesn't recognize most sd cards after issuing a reboot. pcengines said that it may be a problem with the SoC and suggested to not use sd cards for booting.
you have been warned - or you may contribute which currently available cards work nicely. my only working one is a 256mb  - guess how old that one is...

**Comment by jakieu** [comment]
Are you sure about that? My 8GB Sandisk Extreme works just like it should. I could test a 16GB Toshiba if you want. So far, no issues with SD cards..

**Comment by rotanid** [comment]
yes i'm sure, 3 out of the 4 cards i tried weren't detected after a reboot. i had to power-cycle. only the oldest worked, as i wrote. the newest was bought together with the APU2 at varia-store, a 8GB transcend industrial card, TS8GSDHCI+

**Comment by jakieu** [comment]
Did some more tests, Sandisk Ultra 8GB (Class 10 30MB/s) works fine, Toshiba Exceria Pro 16GB (UHS-I, UHS Speed Class 3, THN-M401S0160E2) works fine, too. If i find another card, I could test it too, but with my APU2C4 it just works as it should. Power cycle or reboot, no problems at all. Maybe something's wrong with your board?

**Comment by CRCinAU** [comment]
Also - maybe BIOS version?

**Comment by Strykar** [comment]
@riptidewave93 Oh, but If I set `LED Name to ath9k-phy0` I can toggle the mPCIe slot 2 (which has the radio installed) LED via `defaulton/none`?

@rotanid This thread on the PC Engines forum may be of interest - [APU2 hang on reboot & USB not working](http://pcengines.info/forums/?page=post&id=62158E69-FE7B-46E8-93F2-7885FF6A8C7C&fid=DF5ACB70-99C4-4C61-AFA6-4C0E0DB05B2A) A half decent USB 2.0 drive should wtfpwn any sdcard in the APU2.
Boot the OS off USB3 or SSD, the SD 2.0 spec SoC speeds are going to hold the board back - [sdhci controller: speed?](http://pcengines.info/forums/?page=post&id=A25FEC45-DA7D-4C9B-90BF-57CD5D74E0C7&fid=DF5ACB70-99C4-4C61-AFA6-4C0E0DB05B2A)

**Comment by rotanid** [comment]
@jakieu tried 2 boards.
@CRCinAU newest bios available from pcengines.
@Strykar i already know which other solutions might be better, i just don't want the description to say "boot via SD card possible" if it isn't with every card. so others don't run into the same problem like me, buy the APU2 with an sdcard and notice they have bought the sdcard for nothing.

**Comment by Yurannich** [comment]
Confirm. After few reboots system cant boot from card.

**Comment by Yurannich** [comment]
Any news? I build my universal rig from childrens design kit :)

![dsc_0103](https://cloud.githubusercontent.com/assets/24240671/21174179/3003dd0e-c20e-11e6-8c89-ed67c844cb02.JPG)
![dsc_0104](https://cloud.githubusercontent.com/assets/24240671/21174178/3003729c-c20e-11e6-8b40-75843c53a7f8.JPG)
 Its not complete yet but close

**Comment by CRCinAU** [comment]
Neat.... How do you cool the CPU?

**Comment by Yurannich** [comment]
With this heatsink:
![dsc_0105](https://cloud.githubusercontent.com/assets/24240671/21174617/ceadb766-c210-11e6-8cfe-f3bdce793924.JPG)
And with that schematic: http://hardlock.org.ua/viewtopic.php?f=9&t=12 for fan

**Comment by riptidewave93** [comment]
@Yurannich still waiting for some upstream changes, if I get around to it I will do an updated RFC post sometime after the holidays.

**Comment by rotanid** [comment]
@jakieu bought 2 other cards from sandisk, SDSDB-008G-B35 and SDSDUN-008G-G46 - both seem to work.

**Comment by KyleSanderson** [comment]
I'm running trunk (lede) squashfs on my SSD without issue.

**Comment by riodda** [comment]
Tested the AES-NI engine and seems to be working ok
`root@lede:~# openssl speed aes-128-cbc
The 'numbers' are in 1000s of bytes per second processed.

type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes 
aes-128 cbc      15610.78k    16823.84k    17267.93k    44619.00k    44946.77k
`

`
root@lede:~# openssl speed -evp aes-128-cbc
The 'numbers' are in 1000s of bytes per second processed.

type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes
aes-128-cbc     117902.70k   167208.36k   207137.73k   226139.06k   229578.75k
`

**Comment by riodda** [comment]
I can confirm that squashfs image boots and reboots correctly on Kingston mSATA SMS200S3/30G

**Comment by riodda** [comment]
Would you consider changing the 3g.sh and proto_3g.lua according to this post ? 

https://forum.openwrt.org/viewtopic.php?pid=346619#p346619

**Comment by riptidewave93** [comment]
@riodda That is a change that needs to be done upstream in the LuCI repo at https://github.com/openwrt/luci and the netifd package at https://git.lede-project.org/?p=project/netifd.git;a=summary as those changes are not related to the purpose of this repo.

Because of this, at this time they will not be added.

**Comment by riptidewave93** [comment]
Due to the fact that it will take a while to get this "fully" merged due to some still pending issues (Please see last RFC post at http://lists.infradead.org/pipermail/lede-dev/2016-October/003536.html), I was curious on the thoughts of submitting patches for some of the "complete" parts, such as the LED/GPIO drivers and userspace support. Any thoughts/feedback on this?

**Comment by Strykar** [comment]
Changes rejected from the last RFC:
- add usb-serial-kmod - The goal of this port is to add support for all sold hardware via PC Engines. As they do not sell 2G/3G/4G PCI-E modems, this will not be included.

It would appear some of PC Engine’s  biggest customers use 3G/4G connectivity, see APU3.
I don’t personally, but would still like to see usb-serial-kmod be included given that the board has 1 SIM card slot.
This would at least have the device show up via dmesg/lspci/lsusb, the user can then figure out the rest.

I’ve been using the lede-upstream-port on my APU2 for a few months now with zero issues.
Can't wait for this to be merged!


- S

**Comment by riptidewave93** [comment]
@Strykar If you look closely, that was re-added after a discussion on the RFC. Commit is at https://github.com/riptidewave93/LEDE-APU2/commit/ae7b95dac81f25bf93db842bed2a1001c4684790

Also note that the board profile part of the code can't be merged until the Makefile for the x86 target is configured to support profiles, but that doesn't mean I can't try and get the rest of the code up, such as the LED drivers as mentioned.

**Comment by CRCinAU** [comment]
At the risk of giving a completely unwanted opinion, I'd try to merge as much as possible - then you're left with the "but it only needs this last missing piece" argument instead of "there's the entire stuff ready to be reviewed"... More likely to get forward movement...

**Comment by KyleSanderson** [comment]
Indeed. It's much easier to get pieces in, then the entire subtarget at once.

**Comment by Yurannich** [comment]
A new pfotos from my build:
![dsc_0129](https://cloud.githubusercontent.com/assets/24240671/21479051/0499ebd4-cb84-11e6-8931-c8d6a7198d02.JPG)
![dsc_0130](https://cloud.githubusercontent.com/assets/24240671/21479053/049dc9de-cb84-11e6-88ee-8734f0bc6d7a.JPG)
![dsc_0131](https://cloud.githubusercontent.com/assets/24240671/21479052/049a66b8-cb84-11e6-84a3-90949c9ce140.JPG)
![dsc_0132](https://cloud.githubusercontent.com/assets/24240671/21479050/0499c7c6-cb84-11e6-9d27-fa505c0031f1.JPG)
![dsc_0133](https://cloud.githubusercontent.com/assets/24240671/21479054/04b85c90-cb84-11e6-949e-2d968d75c8ba.JPG)
LEDs driving for J17 PLEASE! :)

**Comment by Yurannich** [comment]
Hi, Chris! Any news?

**Comment by riptidewave93** [comment]
@Yurannich Nothing at this time, but any updates will be posted in the PR at https://github.com/riptidewave93/LEDE-APU2/pull/5

**Comment by riptidewave93** [comment]
Just submitted a handful of commits upstream for review & feedback. Links to said patches can be found in the first post of this thread.

EDIT: LED & GPIO drivers rejected, will work on em later when I get motivation to do so. Anyone else is free to do so in the meantime. Changes requested are at http://lists.infradead.org/pipermail/lede-dev/2017-January/005357.html

**Comment by feckert** [comment]
I have tested the latest lede with apu3a2 and i get the following traceback, if i plug in in a network cable ethx. Any suggestion what this cause the traceback? I have already compiled a x86_generic same result. I have searched the internet and found this issue on the intel page https://communities.intel.com/thread/45060


> 

root@lede:/# /etc/init[   90.361072] igb 0000:02:00.0: Detected Tx Unit Hang                                                            
[   90.361072]   Tx Queue             <2>                                                                                               
[   90.361072]   TDH                  <0>                                                                                               
[   90.361072]   TDT                  <1>                                                                                               
[   90.361072]   next_to_use          <1>                                                                                               
[   90.361072]   next_to_clean        <0>                                                                                               
[   90.361072] buffer_info[next_to_clean]                                                                                               
[   90.361072]   time_stamp           <ffffad55>                                                                                        
[   90.361072]   next_to_watch        <ffff88000001d000>                                                                                
[   90.361072]   jiffies              <ffffadf3>                                                                                        
[   90.361072]   desc.status          <138000>                                                                                          
.[   92.360290] igb 0000:02:00.0: Detected Tx Unit Hang                                                                                 
[   92.360290]   Tx Queue             <2>                                                                                               
[   92.360290]   TDH                  <0>                                                                                               
[   92.360290]   TDT                  <1>                                                                                               
[   92.360290]   next_to_use          <1>                                                                                               
[   92.360290]   next_to_clean        <0>                                                                                               
[   92.360290] buffer_info[next_to_clean]                                                                                               
[   92.360290]   time_stamp           <ffffad55>                                                                                        
[   92.360290]   next_to_watch        <ffff88000001d000>                                                                                
[   92.360290]   jiffies              <ffffaebb>                                                                                        
[   92.360290]   desc.status          <138000>                                                                                          
[   94.360569] igb 0000:02:00.0: Detected Tx Unit Hang                                                                                  
[   94.360569]   Tx Queue             <2>                                                                                               
[   94.360569]   TDH                  <0>                                                                                               
[   94.360569]   TDT                  <1>                                                                                               
[   94.360569]   next_to_use          <1>                                                                                               
[   94.360569]   next_to_clean        <0>                                                                                               
[   94.360569] buffer_info[next_to_clean]                                                                                               
[   94.360569]   time_stamp           <ffffad55>                                                                                        
[   94.360569]   next_to_watch        <ffff88000001d000>                                                                                
[   94.360569]   jiffies              <ffffaf83>                                                                                        
[   94.360569]   desc.status          <138000>                                                                                          
[   96.360772] igb 0000:02:00.0: Detected Tx Unit Hang                                                                                  
[   96.360772]   Tx Queue             <2>                                                                                               
[   96.360772]   TDH                  <0>                                                                                               
[   96.360772]   TDT                  <1>                                                                                               
[   96.360772]   next_to_use          <1>                                                                                               
[   96.360772]   next_to_clean        <0>                                                                                               
[   96.360772] buffer_info[next_to_clean]                                                                                               
[   96.360772]   time_stamp           <ffffad55>                                                                                        
[   96.360772]   next_to_watch        <ffff88000001d000>                                                                                
[   96.360772]   jiffies              <ffffb04b>                                                                                        
[   96.360772]   desc.status          <138000>                                                                                          
[   98.360994] igb 0000:02:00.0: Detected Tx Unit Hang                                                                                  
[   98.360994]   Tx Queue             <2>                                                                                               
[   98.360994]   TDH                  <0>                                                                                               
[   98.360994]   TDT                  <1>                                                                                               
[   98.360994]   next_to_use          <1>                                                                                               
[   98.360994]   next_to_clean        <0>                                                                                               
[   98.360994] buffer_info[next_to_clean]                                                                                               
[   98.360994]   time_stamp           <ffffad55>                                                                                        
[   98.360994]   next_to_watch        <ffff88000001d000>                                                                                
[   98.360994]   jiffies              <ffffb113>                                                                                        
[   98.360994]   desc.status          <138000>                                                                                          
[   98.496475] ------------[ cut here ]------------                                                                                     
[   98.510362] WARNING: CPU: 3 PID: 19 at net/sched/sch_generic.c:306 dev_watchdog+0x1e8/0x200()                                        
[   98.535947] NETDEV WATCHDOG: eth1 (igb): transmit queue 2 timed out                                                                  
[   98.554757] Modules linked in: pppoe ppp_async iptable_nat pppox ppp_generic nf_nat_ipv4 nf_conntrack_ipv6 nf_conntrack_ipv4 ipt_REJi
[   98.729902] CPU: 3 PID: 19 Comm: ksoftirqd/3 Not tainted 4.4.42 #0                                                                   
[   98.748437] Hardware name: PC Engines APU2/APU2, BIOS 4.0.5 01/03/2017                                                               
[   98.768018]  0000000000000000 ffffffff811f3202 ffff880075527d78 ffffffff815e3e29                                                     
[   98.790402]  ffffffff81065b73 0000000000000002 ffff880075527dc8 0000000000000008                                                     
[   98.812789]  0000000000000003 ffff880074bae3e0 ffffffff81065be7 ffffffff815e6c38                                                     
[   98.835154] Call Trace:                                                                                                              
[   98.842529]  [<ffffffff811f3202>] ? dump_stack+0x5c/0x7a                                                                             
[   98.858465]  [<ffffffff81065b73>] ? warn_slowpath_common+0x73/0xa0                                                                   
[   98.877001]  [<ffffffff81065be7>] ? warn_slowpath_fmt+0x47/0x50                                                                      
[   98.894763]  [<ffffffff813caa98>] ? dev_watchdog+0x1e8/0x200                                                                         
[   98.911740]  [<ffffffff813ca8b0>] ? qdisc_rcu_free+0x40/0x40                                                                         
[   98.928726]  [<ffffffff810a3211>] ? call_timer_fn.isra.4+0x11/0x70                                                                   
[   98.947262]  [<ffffffff810a33a0>] ? run_timer_softirq+0x130/0x1f0                                                                    
[   98.965542]  [<ffffffff810684d6>] ? __do_softirq+0xc6/0x1d0                                                                          
[   98.982261]  [<ffffffff81068600>] ? run_ksoftirqd+0x20/0x40                                                                          
[   98.998982]  [<ffffffff8107da74>] ? smpboot_thread_fn+0xf4/0x150                                                                     
[   99.017000]  [<ffffffff8107d980>] ? sort_range+0x20/0x20                                                                             
[   99.032942]  [<ffffffff8107b6e8>] ? kthread+0xb8/0xd0                                                                                
[   99.048101]  [<ffffffff8107b630>] ? kthread_worker_fn+0x100/0x100                                                                    
[   99.066385]  [<ffffffff8147b2cf>] ? ret_from_fork+0x3f/0x70                                                                          
[   99.083101]  [<ffffffff8107b630>] ? kthread_worker_fn+0x100/0x100
   99.101403] ---[ end trace c95caebce5b20f73 ]---                                                                                     
[   99.115315] igb 0000:02:00.0 eth1: Reset adapter                                                                                     
[  103.432359] igb 0000:02:00.0 eth1: igb: eth1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX/TX

**Comment by CRCinAU** [comment]
There's a patch required for the IGB ethernet adapter.

I've just noticed that my nightly lede builds are busted - so I've just resolved that, and a new nightly should appear at:
http://openwrt.crc.id.au/

Try that and see how you go. I use the same patch for my RHEL packaged kernel for the APU2 and all is good.

**Comment by feckert** [comment]
@CRCinAU
Thanks for your reply. I will try your new nightly build tomorrow.

**Comment by feckert** [comment]
@CRCinAU Sorry for the noise but where can i find the patch?

**Comment by CRCinAU** [comment]
That's a great question.

I just noticed that I don't even apply it to my kernel for CentOS / RHEL:
https://xen.crc.id.au/git/?p=kernel-xen;a=blob;f=SPECS/kernel-xen.spec;h=eb48daaa036b34fb18782a6b0b9a60107bde55d6;hb=HEAD

I know we had to in the past.... @riptidewave93 - am I remembering faulty? or was this merged into the upstream kernel.org source - hence we don't need it anymore?

**Comment by riptidewave93** [comment]
@CRCinAU If your referring to the patch at https://github.com/riptidewave93/LEDE-APU2/issues/8 and https://github.com/riptidewave93/LEDE-APU2/blob/lede-upstream-port/overlay/target/linux/x86/patches-4.4/700-igb-spin-lock-fix.patch, as NBD is the creator you would need to reach out to him about getting it merged as I do not feel comfortable submitting his patch in a PR, out of respect.

**Comment by CRCinAU** [comment]
Ahhh that's the one - its coming back to me now.... That fixed the ethernet lockups when using the LED trigger for the onboard LEDs and the netdev trigger source...

I wonder if that's the same issue as above.

**Comment by CRCinAU** [comment]
For what its worth, managed to get my nightly buildserver working again. It pulls the latest from this repo over the top of the nightly lede trunk.

http://openwrt.crc.id.au/

**Comment by riptidewave93** [comment]
Just merged in the PR from @feckert (Thanks again!) as well as changes to move leds-apu2 to a kernel module. If I can get some of you to review both and make sure they work without issue, then I can work on getting em pushed up along with userspace changes.

**Comment by feckert** [comment]
@riptidewave93 if you want then you could try it to get the gpio driver merged into lede. I would suggest to make 2 pull request one for gpio and one for apu leds. I will test your led driver tomorrow. I hope that the drivers will get into to lede soon :-)

@CRCinAU I have already applied the igb patch in this repo. But nothing has changed! Same error. I am unable to get the network up on this device! I even don´t know what the error log means!

**Comment by CRCinAU** [comment]
Out of interest, does the one in my build system work? http://openwrt.crc.id.au

I built one within the last 24 hours.

**Comment by Yurannich** [comment]
Yes, like a charm :)

**Comment by feckert** [comment]
@CRCinAU No it is not working same log output when i plugged in the ethernet cable!
I have a apu3a2. it is optimized for for 3G/LTE modems. But i think it has the same hardware.
Does anyone else have a apu3a2?
Could the Bios be responsible for this issue? I have the version

Coreboot version:

> coreboot build 20170103
> -2064 MB DRAM
> 
> SeaBIOS (version rel-1.9.2.3)

Log Output:

> [   32.961789] igb 0000:01:00.0 eth0: igb: eth0 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX/TX
> [   32.971476] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
> [   35.052212] igb 0000:01:00.0: Detected Tx Unit Hang
> [   35.052212]   Tx Queue             <2>
> [   35.052212]   TDH                  <0>
> [   35.052212]   TDT                  <1>
> [   35.052212]   next_to_use          <1>
> [   35.052212]   next_to_clean        <0>
> [   35.052212] buffer_info[next_to_clean]
> [   35.052212]   time_stamp           <ffff97ac>
> [   35.052212]   next_to_watch        <ffff8800731eb000>
> [   35.052212]   jiffies              <ffff987a>
> [   35.052212]   desc.status          <208000>
> [   35.052228] igb 0000:01:00.0: Detected Tx Unit Hang
> [   35.052228]   Tx Queue             <3>
> [   35.052228]   TDH                  <0>
> [   35.052228]   TDT                  <1>
> [   35.052228]   next_to_use          <1>
> [   35.052228]   next_to_clean        <0>
> [   35.052228] buffer_info[next_to_clean]
> [   35.052228]   time_stamp           <ffff97bf>
> [   35.052228]   next_to_watch        <ffff8800731e7000>
> [   35.052228]   jiffies              <ffff987a>
> [   35.052228]   desc.status          <558000>
> [   37.051313] igb 0000:01:00.0: Detected Tx Unit Hang
> [   37.051313]   Tx Queue             <3>
> [   37.051313]   TDH                  <0>
> [   37.051313]   TDT                  <1>
> [   37.051313]   next_to_use          <1>
> [   37.051313]   next_to_clean        <0>
> [   37.051313] buffer_info[next_to_clean]
> [   37.051313]   time_stamp           <ffff97bf>
> [   37.051313]   next_to_watch        <ffff8800731e7000>
> [   37.051313]   jiffies              <ffff9942>
> [   37.051313]   desc.status          <558000>
> [   37.051327] igb 0000:01:00.0: Detected Tx Unit Hang
> [   37.051327]   Tx Queue             <2>
> [   37.051327]   TDH                  <0>
> [   37.051327]   TDT                  <1>
> [   37.051327]   next_to_use          <1>
> [   37.051327]   next_to_clean        <0>
> [   37.051327] buffer_info[next_to_clean]
> [   37.051327]   time_stamp           <ffff97ac>
> [   37.051327]   next_to_watch        <ffff8800731eb000>
> [   37.051327]   jiffies              <ffff9942>
> [   37.051327]   desc.status          <208000>
> [   38.061188] ------------[ cut here ]------------
> [   38.065861] WARNING: CPU: 2 PID: 0 at net/sched/sch_generic.c:306 dev_watchdog+0x1e8/0x200()
> [   38.074357] NETDEV WATCHDOG: eth0 (igb): transmit queue 2 timed out
> [   38.080649] Modules linked in: pppoe ppp_async iptable_nat ath9k pppox ppp_generic nf_nat_ipv4 nf_conntrack_ipv6 nf_conntrack_ipv4 ie
> [   38.190883] CPU: 2 PID: 0 Comm: swapper/2 Not tainted 4.4.42 #0
> [   38.196832] Hardware name: PC Engines APU2/APU2, BIOS 4.0.5 01/03/2017
> [   38.203386]  0000000000000000 ffffffff811f8492 ffff880077d03e60 ffffffff81625510
> [   38.210919]  ffffffff81065c83 0000000000000002 ffff880077d03eb0 0000000000000008
> [   38.218444]  0000000000000002 ffff880074e2c3e0 ffffffff81065cf7 ffffffff816284d0
> [   38.225966] Call Trace:
> [   38.228441]  <IRQ>  [<ffffffff811f8492>] ? dump_stack+0x5c/0x7a
> [   38.234437]  [<ffffffff81065c83>] ? warn_slowpath_common+0x73/0xa0
> [   38.240641]  [<ffffffff81065cf7>] ? warn_slowpath_fmt+0x47/0x50
> [   38.246590]  [<ffffffff813dd358>] ? dev_watchdog+0x1e8/0x200
> [   38.252278]  [<ffffffff813dd170>] ? qdisc_rcu_free+0x40/0x40
> [   38.257963]  [<ffffffff810a35c1>] ? call_timer_fn.isra.4+0x11/0x70
> [   38.264174]  [<ffffffff810a3750>] ? run_timer_softirq+0x130/0x1f0
> [   38.270294]  [<ffffffff81068606>] ? __do_softirq+0xc6/0x1d0
> [   38.275899]  [<ffffffff8106882f>] ? irq_exit+0x6f/0x80
> [   38.281076]  [<ffffffff8102d063>] ? smp_trace_apic_timer_interrupt+0x63/0x90
> [   38.288157]  [<ffffffff814a197f>] ? apic_timer_interrupt+0x7f/0x90
> [   38.294353]  <EOI>  [<ffffffff81375f9d>] ? cpuidle_enter_state+0x12d/0x1e0
> [   38.301294]  [<ffffffff81375f96>] ? cpuidle_enter_state+0x126/0x1e0
> [   38.307591]  [<ffffffff81091274>] ? cpu_startup_entry+0x1f4/0x250
> [   38.313709]  [<ffffffff8102bb8d>] ? start_secondary+0x12d/0x140
> [   38.319672] ---[ end trace 6688d55a43b7e268 ]---
> [   38.325453] igb 0000:01:00.0 eth0: Reset adapter

**Comment by riptidewave93** [comment]
@feckert sadly I only have APU2 units to test on, but from my understanding the APU2c4 uses i210AT controllers, while the APU3 uses i211AT controllers. This may possibly explain the variance between boards.

I also just submitted the GPIO & LED patches upstream, so please review them (Links in the original issue post) and feel free to share any feedback.

**Comment by rotanid** [comment]
@riptidewave93 according to the following page it is not that easy, as APU2c2 also has i211AT - so this would affect me, too...
http://pcengines.ch/apu2.htm

**Comment by CRCinAU** [comment]
Maybe raise it directly with pcengines.... At this stage, if it isn't the single patch we know about already, anything beyond this would just be a guess on our behalf....

**Comment by feckert** [comment]
@riptidewave93 i have an additional problem with the leds_apu2 driver i have the "PCI Revision ID: 0x42". In the device driver this will be tested and if the revision is greater than 0x40 the gpios and so the leds will not be loaded?

What version of hardware do you have? I think i have a newer hardware. With newer chips!

See http://pcengines.ch/apu2.htm this is the same for the i210AT/i211AT issue that i have?

**Comment by riptidewave93** [comment]
@feckert That's actually strange, as it's based on the chipset version on the southbridge in the SoC. With that said, the driver should still work as it looks for a version above 0x40, per https://github.com/riptidewave93/LEDE-APU2/blob/c2a61880f7a5092d4b4bc2d681e372add8b20606/overlay/package/kernel/leds-apu2/src/leds-apu2.c#L176

Are you getting an actual error returned with the LED driver? If so, can you please share the output?

As for my board, it's a 0x42 as well, and I do not believe this is related to the NIC issue you are seeing.

**Comment by feckert** [comment]
@riptidewave93 sorry for noise about "Tx Unit Hang" of the NIC. I ordered a new board preventive last week and it is working now perfect!! I talked to PC-Engines and the answer was that "maybe the NIC was not initialized during production correctly". 👎 🥇 so this issue is fixed for me

**Comment by rotanid** [comment]
@riptidewave93  thanks for your work again! maybe you could add the igb patch to your list at the top of the issue? just thought "wow only device profile missing" and then saw this igb patch in the repository.

**Comment by CRCinAU** [comment]
the igb patch was never part of this repo - it was part of the upstream LEDE. It was cherry picked for testing - but should be either upstreamed to the main kernel.org or after checking if it isn't already included in LEDE, to discuss with nbd for inclusion.

**Comment by Strykar** [comment]
Has nbd responded/been contacted about this yet? https://github.com/riptidewave93/LEDE-APU2/blob/lede-upstream-port/overlay/target/linux/x86/patches-4.4/700-igb-spin-lock-fix.patch doesn't exist anymore. All the APU2/3 boards except the c4 use the slightly cheaper i211 NICs. Anyone tracking the patches the APU3's dual SIM and GPIO failover require?

APU3 owners would find it useful to have the NICs working even if the dual SIM failover isn't right away.
Would it be acceptable for someone else to submit nbd's patches providing appropriate attribution?

**Comment by CRCinAU** [comment]
None of us have APU3s to play with - as such, patches / pull requests welcome if you've managed to get it working.

**Comment by riptidewave93** [comment]
As for the GPIO part for the APU3, why not just manually export the GPIO? LEDE does not have any web UI function for this in LuCI, and the apu2 driver currently does not have any detection code for the apu3 bios as I do not have a unit to test and develop with.

**Comment by rotanid** [comment]
@Strykar regarding the igb patch, nbd wrote on IRC on thursday the patch shouldn't be necessary anymore, fixed by fixing ledtrig-netdev

**Comment by KyleSanderson** [comment]
@Strykar I'm not certain anyone should get anything beyond the APU2C4; anything else comes with the cheaper NICs; including the entirety of the APU3.

**Comment by obsy** [comment]
I have access to apu3. This board is recognized by bios as apu2 (ver 20160519)

**Comment by jakieu** [comment]
Do you have any plans to switch to Kernel 4.9? I think it's slowly rolling into LEDE Master ;-)

**Comment by riptidewave93** [comment]
@jakieu Waiting on the current pending PR to get merged before I update this repo, at which point, only the device profile would be left.

**Comment by riptidewave93** [comment]
OK, repo is up to date and no longer runs on targeted git commits as the "core" stuff is all upstream. :)

**Comment by jakieu** [comment]
Do you have any recommendations what should (not) be enabled in kernelconfig to support cryptographic coprocessor (AES-NI) for OpenVPN?

**Comment by riptidewave93** [comment]
@jakieu That should already be enabled upstream in LEDE per commit https://github.com/lede-project/source/commit/301301da2b2998177d4c34521f933c828972a985

**Comment by riptidewave93** [comment]
PR for the Profile added at https://github.com/lede-project/source/pull/883

**Comment by riptidewave93** [comment]
Going to close this issue out as:

1. The last PR was pushed and rejected, but is being discussed
2. This thread is getting too long and somewhat offtopic.

For any new issues related to unmerged code, please create an issue. For anything that was merged, please use official discussion channels (IRC/Mailing list/Bugtracker)

**Comment by KyleSanderson** [comment]
Damn shame man. It's too bad @nbd seems this as a burden then an absolute
addition.
